packages/babel-parser/src/parser/expression.js

```diff
@@ -2183,6 +2183,7 @@ export default class ExpressionParser extends LValParser {
   isAwaitAllowed(): boolean {
     if (this.scope.inFunction) return this.scope.inAsync;
     if (this.options.allowAwaitOutsideFunction) return true;
+    if (this.hasPlugin("topLevelAwait")) return this.inModule;
```

Comment:
If top-level await is only valid in modules, should seeing one set `sawUnambiguousESM`? In other words, how should this behave with a `sourceType` of `unambiguous`? Currently, parsing `await 0` with `unambiguous` yields an AST with `sourceType` of `script`, even though if you'd specified `script` it would not have parsed:

```
> require('./').parse('await 0', { plugins: ['topLevelAwait'], sourceType: 'unambiguous' }).program.sourceType
'script'
> require('./').parse('await 0', { plugins: ['topLevelAwait'], sourceType: 'script' }).program.sourceType
Thrown:
{ [SyntaxError: Unexpected token, expected ";" (1:6)
] pos: 6, loc: Position { line: 1, column: 6 } }
```

Reply:
This is tricky because, for exmaple, this is both a valid script and a valid module:
```js
await
0
```

I can set it to a module when `await` the next invalid token is on the same line.